### PR TITLE
ConcurrentModificationException when late meter filters are added

### LINKED ISSUE #5489: ConcurrentModificationException when late meter filters are added
**Describe the bug**
In Keycloak, we're facing an issue (https://github.com/keycloak/keycloak/issues/33246) that the server is unable to start when metrics are enabled - we use Quarkus under the hood with the Quarkus Micrometer extension. 

It started to happen after the Micrometer 1.13.x upgrade on the Quarkus side.
It happens only sometimes, and is reproducible after multiple start/stops of Keycloak (Quarkus-based) server.

The issue is probably touching the addition of meter filters after the creation of Meters described here: https://github.com/micrometer-metrics/micrometer/issues/4920

The related log:
```
2024-09-24 15:24:07,330 INFO  [org.jgroups.protocols.FD_SOCK2] (ForkJoinPool.commonPool-worker-1) server listening on *.34764
2024-09-24 15:24:07,389 INFO  [org.jgroups.protocols.pbcast.GMS] (ForkJoinPool.commonPool-worker-1) mabartos-12454: no members discovered after 55 ms: creating cluster as coordinator
2024-09-24 15:24:07,406 INFO  [org.infinispan.CLUSTER] (ForkJoinPool.commonPool-worker-1) ISPN000094: Received new cluster view for channel ISPN: [mabartos-12454|0] (1) [mabartos-12454]
2024-09-24 15:24:07,489 INFO  [org.infinispan.CLUSTER] (ForkJoinPool.commonPool-worker-1) ISPN000079: Channel `ISPN` local address is `mabartos-12454`, physical addresses are `[10.202.149.163:50300]`
2024-09-24 15:24:07,901 DEBUG [io.micrometer.common.util.internal.logging.InternalLoggerFactory] (main) Using SLF4J as the default logging framework
2024-09-24 15:24:07,903 DEBUG [io.quarkus.micrometer.runtime.MicrometerRecorder] (main) Discovered global MeterFilters : [io.micrometer.core.instrument.config.MeterFilter$9@15311af2, io.micrometer.core.instrument.config.MeterFilter$9@796434a3, io.micrometer.core.instrument.config.MeterFilter$9@1626612d, io.micrometer.core.instrument.config.MeterFilter$9@542aba71]
2024-09-24 15:24:07,904 DEBUG [io.quarkus.micrometer.runtime.MicrometerRecorder] (main) Discovered MeterFilters for class io.micrometer.prometheus.PrometheusMeterRegistry: []
2024-09-24 15:24:07,904 DEBUG [io.quarkus.micrometer.runtime.MicrometerRecorder] (main) Discovered global MeterRegistryCustomizer : []
2024-09-24 15:24:07,905 DEBUG [io.quarkus.micrometer.runtime.MicrometerRecorder] (main) Configuring Micrometer registries : [class io.micrometer.prometheus.PrometheusMeterRegistry]
2024-09-24 15:24:07,905 DEBUG [io.quarkus.micrometer.runtime.MicrometerRecorder] (main) Discovered MeterRegistryCustomizer for class io.micrometer.prometheus.PrometheusMeterRegistry: []
2024-09-24 15:24:07,907 INFO  [com.arjuna.ats.jbossatx] (main) ARJUNA032014: Stopping transaction recovery manager
ERROR: Failed to start server in (production) mode
Error details:
java.util.ConcurrentModificationException
	at java.base/java.util.HashMap$HashIterator.nextNode(HashMap.java:1605)
	at java.base/java.util.HashMap$KeyIterator.next(HashMap.java:1628)
	at java.base/java.util.AbstractCollection.addAll(AbstractCollection.java:337)
	at io.micrometer.core.instrument.MeterRegistry$Config.meterFilter(MeterRegistry.java:832)
	at io.quarkus.micrometer.runtime.MicrometerRecorder.applyMeterFilters(MicrometerRecorder.java:233)
	at io.quarkus.micrometer.runtime.MicrometerRecorder.configureRegistries(MicrometerRecorder.java:89)
	at io.quarkus.deployment.steps.MicrometerProcessor$configureRegistry1996010268.deploy_0(Unknown Source)
	at io.quarkus.deployment.steps.MicrometerProcessor$configureRegistry1996010268.deploy(Unknown Source)
	at io.quarkus.runner.ApplicationImpl.doStart(Unknown Source)
	at io.quarkus.runtime.Application.start(Application.java:101)
	at io.quarkus.runtime.ApplicationLifecycleManager.run(ApplicationLifecycleManager.java:119)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:71)
	at org.keycloak.quarkus.runtime.KeycloakMain.start(KeycloakMain.java:147)
	at org.keycloak.quarkus.runtime.cli.command.AbstractStartCommand.run(AbstractStartCommand.java:51)
	at picocli.CommandLine.executeUserObject(CommandLine.java:2030)
	at picocli.CommandLine.access$1500(CommandLine.java:148)
	at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2465)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2457)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2419)
	at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2277)
	at picocli.CommandLine$RunLast.execute(CommandLine.java:2421)
	at picocli.CommandLine.execute(CommandLine.java:2174)
	at org.keycloak.quarkus.runtime.cli.Picocli.parseAndRun(Picocli.java:138)
	at org.keycloak.quarkus.runtime.KeycloakMain.main(KeycloakMain.java:107)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at io.quarkus.bootstrap.runner.QuarkusEntryPoint.doRun(QuarkusEntryPoint.java:62)
	at io.quarkus.bootstrap.runner.QuarkusEntryPoint.main(QuarkusEntryPoint.java:33)

```

**Environment**

- Quarkus version - 3.14.4 
- Micrometer version - 1.13.3
 - Micrometer registry - Prometheus
 - OS: Linux 6.10.6-100.fc39.x86_64 SMP PREEMPT_DYNAMIC GNU/Linux
 - Java version: openjdk 21.0.3 2024-04-16 LTS

**To Reproduce**
How to reproduce the bug:

- Start Keycloak server with enabled Quarkus micrometer extension sequentially multiple times

 f.e. this test case https://github.com/keycloak/keycloak/blob/main/quarkus/tests/integration/src/test/java/org/keycloak/it/cli/dist/StartCommandDistTest.java#L134

**Expected behavior**
No `ConcurrentModificationException`

**Additional context**
It seems the problem is lack of synchronization

### PR DESCRIPTION
**Pull request by mabartos**
- Fixes #5489

@lenin-jaganathan You've already mentioned that it should be synchronized, but no action were taken (https://github.com/micrometer-metrics/micrometer/pull/4857/commits/778f400174e54935a5948f17c283713edb912258#r1555315425)

Is it ok to lock it with the `meterMapLock`, or rather use a concurrent collection?

### CODE REVIEW

**Review by brunobat — APPROVED** [review]

**Comment by lenin-jaganathan** [comment]
I wouldn't recommend a `Collections.synchronizedSet()` if that is what you mean by "concurrent collection". We use `contains` on every interaction with the builder which could cause slowness.

We can use `meterMapLock` (but it was not originally meant for this map) but we should guard all the writes atleast. I also see that this collection is declared as not "Thread safe".

**Comment by mabartos** [comment]
@lenin-jaganathan Thanks for the comment.

> I wouldn't recommend a Collections.synchronizedSet() if that is what you mean by "concurrent collection". We use contains on every interaction with the builder which could cause slowness.

Ok.

> We can use meterMapLock (but it was not originally meant for this map) but we should guard all the writes atleast. I also see that this collection is declared as not "Thread safe".

All the writes/removes should be guarded now. 

@lenin-jaganathan Let me know, if there's a better approach. But resolving this bug is crucial for us as it would need to be propagated to Quarkus as well, so it'd be great to prioritize this issue.

**Comment by mabartos** [comment]
The CI `build-jdk11` failed due to a test failure related to some timing issue after providing `synchronized()` to `unmarkStaleId()` method. Not sure if it is related to this PR, or just the test is unstable.

**Comment by shakuzen** [comment]
I re-ran the jdk11 job and it passed. I'm not sure if the failures were related to the change or not.
Note I'm holding off reviewing the specific changes in this PR and running benchmarks until I can get a concurrency test that reproduces the failure this is fixing. We want to make sure we've really fixed it and that it stays fixed through future changes. I'm out of time to continue working on that today but I'll get back to it tomorrow.

**Comment by shakuzen** [comment]
I was finally able to figure out a test that reproduces the CME. It wasn't obvious to me at first, but the CME is being thrown because `preFilterIdToMeterMap` is being modified (a new entry added) while it is being iterated over. The iteration is implicit in the foreach call in `addAll` that's being passed the keySet of `preFilterIdToMeterMap`. With a test that's hopefully reproducing the same issue as the one reported, we can work on the best solution for it. Unfortunately I'm a bit under the weather today, but if I'm feeling well enough tomorrow I'll continue with this work then.

Code:

```java
@JCStressTest
@Outcome(id = "true", expect = Expect.ACCEPTABLE, desc = "No exception")
@Outcome(expect = Expect.FORBIDDEN, desc = "ConcurrentModificationException thrown")
@State
public static class ConfigureLateMeterFilterWithNewMeterRegister {
    MeterRegistry registry = new SimpleMeterRegistry();

    public ConfigureLateMeterFilterWithNewMeterRegister() {
        // need the registry to not be empty
        registry.counter("c1");
    }

    @Actor
    public void actor1() {
        // creates new meter to add to preMap, updating the modCount
        registry.counter("c2");
    }

    @Actor
    public void actor2(Z_Result r) {
        try {
            // adds all the preMap keys to the staleIds, iterating over preMap keys
            registry.config().commonTags("common2", "tag2");
            r.r1 = true;
        } catch (ConcurrentModificationException e) {
            r.r1 = false;
        }
    }
}
```

Results:

```
  Results across all configurations:

  RESULT     SAMPLES     FREQ      EXPECT  DESCRIPTION
   false      57,671    0.19%   Forbidden  ConcurrentModificationException thrown
    true  30,863,033   99.81%  Acceptable  No exception

```

**Comment by mabartos** [comment]
@shakuzen After applying these changes, the test results are as follows:

```
[OK] io.micrometer.concurrencytests.MeterRegistryConcurrencyTest.ConfigureLateMeterFilterWithNewMeterRegister

  Results across all configurations:

  RESULT    SAMPLES     FREQ      EXPECT  DESCRIPTION
    true  8,239,104  100.00%  Acceptable  No exception

```

Let me know if I can help somehow.

File: micrometer-core/src/main/java/io/micrometer/core/instrument/MeterRegistry.java [hunk 1]
```diff
@@ -692,7 +695,12 @@ private boolean isStaleId(Id originalId) {
      * @return {@code true} if the id is stale
      */
     private boolean unmarkStaleId(Id originalId) {
-        return !stalePreFilterIds.isEmpty() && stalePreFilterIds.remove(originalId);
+        if (!stalePreFilterIds.isEmpty()) {
+            synchronized (meterMapLock) {
```
Comment (shakuzen) [inline micrometer-core/src/main/java/io/micrometer/core/instrument/MeterRegistry.java:699]: This isn't needed for fixing the CME and it's not clear to me what scenario this improves (it also doesn't seem to hurt anything either, though). I am leaning toward not including this change unless we have a concrete idea on that.

**Review by shakuzen — COMMENTED** [review]
I left a comment about one of the changes. I have some polish to make for the concurrency test, and we will want to rebase on the `1.13.x` branch so we can release the fix in the next maintenance release, but I can take care of both of these things before merging.

File: micrometer-core/src/main/java/io/micrometer/core/instrument/MeterRegistry.java (hunk 1 above)
Comment (mabartos) [reply micrometer-core/src/main/java/io/micrometer/core/instrument/MeterRegistry.java:699]: @shakuzen You're right it's not needed for fixing the actual issue, but it was mainly provided after @lenin-jaganathan comment that all writes(write ops) should be covered as we also stated in the comment of the collection. 

But I'll update it.

**Comment by mabartos** [comment]
@shakuzen I updated the code and squashed commits.

Created backport for 1.13.x branch: https://github.com/micrometer-metrics/micrometer/pull/5496

Not sure how you handle backports, so let me know if I should manage it differently. 

We need to have the new patch 1.13.x version ASAP to include it in Quarkus and other project. 

Thanks for your understanding.

**Comment by brunobat** [comment]
> 
> We need to have the new patch 1.13.x version ASAP to include it in Quarkus and other project.
> 

Not sure they can do that...
If there are only fixes in the next patch release we can roll forward and backport to the Quarkus LTS branch.

**Comment by shakuzen** [comment]
> Not sure how you handle backports, so let me know if I should manage it differently.

No worries. We merge to the oldest applicable maintenance branch and then merge forward from there. I'll take care of it from here.

**Comment by mabartos** [comment]
> No worries. We merge to the oldest applicable maintenance branch and then merge forward from there. I'll take care of it from here.

@shakuzen Ok, thank you!

**Pull request merged**